acme/client.go

```diff
@@ -100,6 +100,24 @@ func NewClient(caDirURL string, user User, keyBits int) (*Client, error) {
 	return &Client{directory: dir, user: user, jws: jws, keyBits: keyBits, solvers: solvers}, nil
 }
 
+// SetChallengeProvider specifies a custom provider that will make the solution available
+func (c *Client) SetChallengeProvider(challenge string, p ChallengeProvider) {
+	switch challenge {
+	case "http-01":
+		if chlng, ok := c.solvers[challenge]; ok {
+			chlng.(*httpChallenge).provider = p
+		}
+	case "tls-sni-01":
+		if chlng, ok := c.solvers[challenge]; ok {
+			chlng.(*tlsSNIChallenge).provider = p
+		}
+	case "dns-01":
+		if chlng, ok := c.solvers[challenge]; ok {
```

Comment:
@xenolf I wasn't sure how you intended `dnsChallenge` to get exposed, but I think this works. If you do `SetChallengeProvider(acme.DNS01, ....)` it will now wrap the provider in `dnsChallenge` as needed.

resolved.

note: i don't think this same situation applies to http01 & tls-sni-01 as well since they start out as present in `Client`. If my assumption is wrong i can make the same change there.